File: Miscellaneous/Other/Review/Stage 1/Review 4/prim/prim.py
from typing import List


def prim(edges: List[List[int]], points: int) -> tuple[int, List[List[int]]]:
    path: List[List[int]] = []
    cost = 0

    count = points - 1
    visited = [False] * points
    record: List[List[int]] = []

    # init
    visited[0] = True
    for item in edges:
        if item[0] == 0 or item[1] == 0:
            record.append(item)

    while count > 0:
        record.sort(key=lambda x: x[2])
        v, w, weight = record.pop(0)

        if not visited[v] or not visited[w]:
            cost += weight
            path.append([v, w])
            count -= 1
            next_node = v if not visited[v] else w
            visited[next_node] = True

            for item in edges:
                if (item[0] == next_node and not visited[item[1]]) or (
                    item[1] == next_node and not visited[item[0]]
                ):
                    record.append(item)

    return [cost, path]


# test
res = prim(
    [
        [0, 1, 4],
        [0, 2, 13],
        [0, 3, 7],
        [0, 4, 7],
        [1, 2, 9],
        [1, 3, 3],
        [1, 4, 7],
        [2, 3, 10],
        [2, 4, 14],
        [3, 4, 4],
    ],
    5,
)
print(res)

File: Miscellaneous/Other/Review/Stage 1/Review 1/prim/prim.ts
export function prim(graph: number[][], vertices: number): [number, number[][]] {
  let cost = 0;
  const path: number[][] = [];
  let count = vertices - 1;

  const visited = new Array(vertices).fill(false);
  const record_array: number[][] = [];

  // init
  for (const item of graph) {
    if (item[0] === 0 || item[1] === 0) {
      record_array.push(item);
    }
  }
  visited[0] = true;

  while (record_array.length && count > 0) {
    record_array.sort((a, b) => a[2] - b[2]);

    const [v, w, weight] = record_array.shift()!;

    if (!visited[v] || !visited[w]) {
      cost = cost + weight;
      path.push([v, w]);
      const new_vertex = visited[v] ? w : v;
      visited[new_vertex] = true;
      count--;

      for (const item of graph) {
        if ((item[0] === new_vertex && !visited[item[1]]) || (item[1] === new_vertex && !visited[item[0]])) {
          record_array.push(item);
        }
      }
    }
  }

  return [cost, path];
}

// test
const res = prim(
  [
    [0, 1, 4],
    [0, 2, 13],
    [0, 3, 7],
    [0, 4, 7],
    [1, 2, 9],
    [1, 3, 3],
    [1, 4, 7],
    [2, 3, 10],
    [2, 4, 14],
    [3, 4, 4],
  ],
  5
);
console.log(res);
